Firestore/Swift/Source/PropertyWrapper/FirestoreQuery.swift

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2021 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+import SwiftUI
+
+public enum FIRPredicate {
+  case isEqualTo(_ lhs: String, _ rhs: Any)
+
+  case isIn(_ lhs: String, _ rhs: [Any])
+  case isNotIn(_ lhs: String, _ rhs: [Any])
+
+  case arrayContains(_ lhs: String, _ rhs: Any)
+  case arrayContainsAny(_ lhs: String, _ rhs: [Any])
+
+  case isLessThan(_ lhs: String, _ rhs: Any)
+  case isGreaterThan(_ lhs: String, _ rhs: Any)
+
+  case isLessThanOrEqualTo(_ lhs: String, _ rhs: Any)
+  case isGreaterThanOrEqualTo(_ lhs: String, _ rhs: Any)
+
+  case orderBy(_ lhs: String, _ rhs: Bool)
+
+  case limitTo(_ lhs: Int)
+
+  /*
+   Factory methods to expose the underlying enum cases with a nicer development experience and improved semantics.
+   */
+  public static func whereField(_ lhs: String, isEqualTo rhs: Any) -> FIRPredicate {
+    .isEqualTo(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isIn rhs: [Any]) -> FIRPredicate {
+    .isIn(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isNotIn rhs: [Any]) -> FIRPredicate {
+    .isNotIn(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, arrayContains rhs: Any) -> FIRPredicate {
+    .arrayContains(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, arrayContainsAny rhs: [Any]) -> FIRPredicate {
+    .arrayContainsAny(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isLessThan rhs: Any) -> FIRPredicate {
+    .isLessThan(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isGreaterThan rhs: Any) -> FIRPredicate {
+    .isGreaterThan(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isLessThanOrEqualTo rhs: Any) -> FIRPredicate {
+    .isLessThanOrEqualTo(lhs, rhs)
+  }
+
+  public static func whereField(_ lhs: String, isGreaterThanOrEqualTo rhs: Any) -> FIRPredicate {
+    .isGreaterThanOrEqualTo(lhs, rhs)
+  }
+
+  public static func order(by lhs: String, descending rhs: Bool = false) -> FIRPredicate {
+    .orderBy(lhs, rhs)
+  }
+
+  public static func limit(to lhs: Int) -> FIRPredicate {
+    .limitTo(lhs)
+  }
+}
+
+@available(iOS 13.0, *)
+internal class QueryStore<T: Decodable>: ObservableObject {
```

Comment:
Not sure `QueryStore` is a good choice, as the term _Store_ has such an overloaded meaning already in the context of Firestore.

@ryanwilson @paulb777 @schmidt-sebastian WDYT?